src/qpx_mmc_css.cpp

```diff
@@ -0,0 +1,1909 @@
+    i_fd = open(drive->media.dvdcss.psz_cachefile, O_RDONLY);
+    b_cache = 1;
+
+    if (i_fd >= 0) {
+      char psz_key[DVD_KEY_SIZE * 3];
+      unsigned int k0, k1, k2, k3, k4;
+
+      psz_key[DVD_KEY_SIZE * 3 - 1] = '\0';
+
+      if (read(i_fd, psz_key, DVD_KEY_SIZE * 3 - 1) == DVD_KEY_SIZE * 3 - 1 &&
+          sscanf(psz_key, "%x:%x:%x:%x:%x", &k0, &k1, &k2, &k3, &k4) == 5) {
+        p_title_key[0] = k0;
+        p_title_key[1] = k1;
+        p_title_key[2] = k2;
+        p_title_key[3] = k3;
+        p_title_key[4] = k4;
+        css_printkey((char *)"CSS: title key found in cache ", p_title_key);
+
+        /* Don't try to save it again */
+        b_cache = 0;
+        i_ret = 1;
+      }
+
+      close(i_fd);
+    }
+  }
+#endif
+
+  /* Crack or decrypt CSS title key for current VTS */
+  if (i_ret < 0) {
+    i_ret = css_titlekey(drive, lba, p_title_key);
+
+    if (i_ret < 0) {
+      printf("CSS: fatal error in vts css key\n");
+      return i_ret;
+    }
+
+    if (i_ret == 0) {
+      printf("CSS: unencrypted title\n");
+      /* We cache this anyway, so we don't need to check again. */
+    }
+  }
+
+#if (DVDCSS_KEY_CACHE > 0)
+  /* Key is valid, we store it on disk. */
+  if (drive->media.dvdcss.psz_cachefile[0] && b_cache) {
+    i_fd = open(drive->media.dvdcss.psz_cachefile, O_RDWR | O_CREAT, 0644);
+    if (i_fd >= 0) {
+      char psz_key[DVD_KEY_SIZE * 3 + 2];
+
+      sprintf(psz_key, "%02x:%02x:%02x:%02x:%02x\r\n", p_title_key[0],
+              p_title_key[1], p_title_key[2], p_title_key[3], p_title_key[4]);
+
+      write(i_fd, psz_key, DVD_KEY_SIZE * 3 + 1);
+      close(i_fd);
+    }
+  }
+#endif
+
+  /* Find our spot in the list */
+  p_newtitle = NULL;
+  p_title = drive->media.dvdcss.p_titles;
+  while ((p_title != NULL) && (p_title->i_startlb < lba)) {
+    p_newtitle = p_title;
+    p_title = p_title->p_next;
+  }
+
+  /* Save the found title */
+  p_title = p_newtitle;
+
+  /* Write in the new title and its key */
+  p_newtitle = (dvd_title_t *)malloc(sizeof(dvd_title_t));
+  p_newtitle->i_startlb = lba;
+  memcpy(p_newtitle->p_key, p_title_key, DVD_KEY_SIZE);
+
+  /* Link it at the head of the (possibly empty) list */
+  if (p_title == NULL) {
+    p_newtitle->p_next = drive->media.dvdcss.p_titles;
+    drive->media.dvdcss.p_titles = p_newtitle;
+  }
+  /* Link the new title inside the list */
+  else {
+    p_newtitle->p_next = p_title->p_next;
+    p_title->p_next = p_newtitle;
+  }
+
+  memcpy(drive->media.dvdcss.TK, p_title_key, DVD_KEY_SIZE);
+  return 0;
+}
+
+/*****************************************************************************
+ * _dvdcss_titlekey: get title key.
+ *****************************************************************************/
+int css_titlekey(drive_info *drive, int lba, dvd_key_t p_title_key) {
+  //    static unsigned char p_garbage[ DVDCSS_BLOCK_SIZE ];  /* we never read
+  //    it back */
+  unsigned char p_key[DVD_KEY_SIZE];
+  int i, i_ret = 0;
```

Comment:
Codacy found an issue: [The scope of the variable 'i' can be reduced.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)